dev/MRTCore/mrt/Microsoft.Windows.ApplicationModel.Resources/src/ResourceContext.cpp

```diff
@@ -1,15 +1,76 @@
-﻿// Copyright (c) Microsoft Corporation and Contributors.
+// Copyright (c) Microsoft Corporation and Contributors.
 // Licensed under the MIT License.
 
 #include "pch.h"
 #include "ResourceContext.h"
 #include "ResourceContext.g.cpp"
 #include "winrt/Windows.Globalization.h"
+#include <wil\resource.h>
 
 const wchar_t c_languageQualifierName[] = L"Language";
 
 namespace winrt::Microsoft::Windows::ApplicationModel::Resources::implementation
 {
+
+ResourceContext::~ResourceContext()
+{
+    winrt::slim_lock_guard guard {m_lock};
+    MrmDestroyResourceContext(m_resourceContext);
+}
+
+
+winrt::Windows::Foundation::Collections::IMap<hstring, hstring> ResourceContext::QualifierValues()
+{
+    m_lock.lock_shared();
```

Comment:
>m_lock.lock_shared(); [](http://example.com/codeflow?start=4&length=21)

is this needed?

Reply:
Unless we move initialization to the ctor, in which case we loose the benifits of lazy initialization without this guard we're at risk of having a race.

Reply:
would this be sufficient?
if (!IsInitialized())
    {
        m_lock.lock();

        if (!IsInitialized())
        InitializeQualifierValueMap();

        m_lock.unlock();

    }

    return m_qualifierValueMap;

Reply:
no, becasue you can't garantee that !IsInitialized() is atomic (I mean, even if we could make the check with a single operation, instead of calling a function, we would also have to make it atomic becasue of caches, etc.). Since the operation isn't atomic, the only way to avoid potential races is to take a lock on it.

Reply:
then can you do the lock inside IsInitialized()? also create lock object, and there is no need to introduce wil.

Reply:
what is the issue with wil?

Reply:
All this is gone, now using call_once.